### PR TITLE
[Benchmark] Add benchmark script for CPU offloading 

### PR DESCRIPTION
**Pull request by ApostaC**
_Note: This PR is part of the big CPU offloading PR #10874 -- this PR contains a benchmark script of "long document QA" use case, to test the effectiveness of CPU offloading_

## About the workload
The workload has N different long documents. It will shuffle the documents and query the serving engine multiple times, where each time it has N queries + each query has one of the documents as the prefix. 

## Example usage:

The following command runs the vLLM with 50GB CPU memory for offloading. 
The workload samples 8 different prompts with a default input length of 20K tokens, then replicates each prompt 2 times.

```bash
    python benchmark_long_document_qa.py \
        --model meta-llama/Llama-2-7b-chat-hf \
        --enable-prefix-caching \
        --block-allocator CpuOffloadingBlockAllocator \
        --num-documents 8 \
        --repeat-count 2 \
        --cpu-memory-gb 50
```

### CODE REVIEW

File: benchmarks/benchmark_long_document_qa.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
+"""
+Benchmark the efficiency of prefix caching.
```
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa.py:2]:
```suggestion
Benchmark long document QA.
```

File: benchmarks/benchmark_long_document_qa.py
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa.py:1]: Better to name it `benchmark_long_document_qa_throughput.py` to indicate that this is an offline benchmark.

File: benchmarks/benchmark_long_document_qa.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+import random
+import time
+
+import torch
+
+from vllm import LLM, SamplingParams
+from vllm.utils import FlexibleArgumentParser
+
+execution_times = {}
```
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa.py:56]: Add comment to explain its format and where to use it.

File: benchmarks/benchmark_long_document_qa.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+
+def build_result_dict(start_time, end_time, *args):
```
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa.py:59]: Add docstring to all functions in this file.

File: benchmarks/benchmark_long_document_qa.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    prompts = [
+        str(i) + ' '.join(['hi'] * args.document_length)
+        for i in range(args.num_documents)
+    ]
+
+    preemption_mode = ""
+    if args.block_allocator == "CpuOffloadingBlockAllocator":
+        preemption_mode = "recompute"
+    else:
+        preemption_mode = "swap"
```
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa.py:146]: I don't think we should set preemption mode in this script. It should be handled by the engine args directly.

File: benchmarks/benchmark_long_document_qa.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+              trust_remote_code=True,
+              enforce_eager=True,
+              tensor_parallel_size=args.tensor_parallel_size,
+              enable_prefix_caching=args.enable_prefix_caching,
+              block_allocator=args.block_allocator,
+              preemption_mode=preemption_mode,
+              swap_space=args.cpu_memory_gb,
+              enable_chunked_prefill=False,
+              gpu_memory_utilization=args.gpu_memory_utilization,
+              max_model_len=30000)
```
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa.py:159]: Can we use `EngineArgs` instead of manually passing arguments? You can refer to `benchmarks/benchmark_throughput.py:L164`

File: benchmarks/benchmark_long_document_qa.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    sampling_params = SamplingParams(temperature=0, max_tokens=args.output_len)
+
+    prompts = repeat_prompts(prompts, args.repeat_count)
+
+    print("------warm up------")
+    test_long_document_qa(
+        llm=llm,
+        prompts=prompts,
+        sampling_params=sampling_params,
+    )
```
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa.py:170]: Will warmup make the result too optimistic because the prefix can potentially be cached?

File: benchmarks/benchmark_long_document_qa.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+    random.shuffle(prompts)
```
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa.py:172]: In the case of long context, the order of prompts will significantly affect the results. I'd suggest instead of random shuffling, we provide several modes for users to select. For example:
- In-order: All repeat documents are consecutive (best case with the highest cache hit).
- Round Robin: Worst case with the lowest cache hit.
- Random: Shuffle by random seed. Should have a warning saying that it is recommended to run multiple times with different random seeds to get a more general result.
- (some other orders you can think of

File: benchmarks/benchmark_long_document_qa.py (hunk 1 above)
Comment (ApostaC) [reply benchmarks/benchmark_long_document_qa.py]: Fixed in the new commit

File: benchmarks/benchmark_long_document_qa.py
Comment (ApostaC) [reply benchmarks/benchmark_long_document_qa.py:1]: Done

File: benchmarks/benchmark_long_document_qa.py (hunk 2 above)
Comment (ApostaC) [reply benchmarks/benchmark_long_document_qa.py]: For simplicity, I removed all the functions and global variables related to profiling.

File: benchmarks/benchmark_long_document_qa.py (hunk 3 above)
Comment (ApostaC) [reply benchmarks/benchmark_long_document_qa.py]: Done

File: benchmarks/benchmark_long_document_qa.py (hunk 4 above)
Comment (ApostaC) [reply benchmarks/benchmark_long_document_qa.py]: Changed the arg parser to use EngineArgs

File: benchmarks/benchmark_long_document_qa.py (hunk 5 above)
Comment (ApostaC) [reply benchmarks/benchmark_long_document_qa.py]: Done

File: benchmarks/benchmark_long_document_qa.py (hunk 6 above)
Comment (ApostaC) [reply benchmarks/benchmark_long_document_qa.py]: Changed the code to make warm-up prompts use different prefixes.

File: benchmarks/benchmark_long_document_qa.py (hunk 7 above)
Comment (ApostaC) [reply benchmarks/benchmark_long_document_qa.py]: Now it supports 3 different modes:
- "random": randomly shuffle the repeated prompts
- "tile": the entire prompt list is repeated in sequence. 
- "interleave": each prompt is repeated consecutively before moving to the next element.

**Comment by ApostaC** [comment]
Hey @comaniac , thanks for all the comments. I've addressed them and the PR is ready for review again.

File: benchmarks/benchmark_long_document_qa_throughput.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+from vllm import LLM, SamplingParams
+from vllm.engine.arg_utils import EngineArgs
+from vllm.utils import FlexibleArgumentParser
+
+
+def test_long_document_qa(llm=None, sampling_params=None, prompts=None):
+    """
+    Test long document QA with the given prompts and sampling parameters.
+    Print the time cost processing all the prompts.
+    """
```
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa_throughput.py:55]: Please follow the docstring style:

```python
    """
    Test long document QA with the given prompts and sampling parameters.
    Print the time cost processing all the prompts.

    Args:
        llm: ...
        sampling_params: ...
        prompts: ...
    """
```

**Review by comaniac — APPROVED** [review]
Otherwise LGTM. Thanks!

File: benchmarks/benchmark_long_document_qa_throughput.py (hunk 8 above)
Comment (ApostaC) [reply benchmarks/benchmark_long_document_qa_throughput.py]: Now fixed. Should be ready to merge now
